hpcbench/campaign.py

```diff
@@ -225,18 +225,60 @@ def _expand_tag_pattern(cls, pattern):
             if mode == 'match':
                 pattern[mode] = re.compile(pattern[mode])
             elif mode == 'nodes':
-                pattern[mode] = NetworkConfig._expand_nodes(pattern[mode])
+                pattern[mode] = cls._expand_nodes(pattern[mode])
+            elif mode == 'tags':
+                pass  # don't fail but ignore tags
             else:
                 raise Exception('Unknown tag association pattern: %s',
                                 mode)
 
-    def _expand_tag(self, tag):
-        config = self.network.tags[tag]
-        if isinstance(config, dict):
-            config = [config]
-            self.network.tags[tag] = config
-        for pattern in config:
-            self._expand_tag_pattern(pattern)
+    @classmethod
+    def _is_leaf(cls, config):
+        # returns True if in none of the modes and patterns is 'tags'
+        return reduce(lambda x, y: x & y,
+                      map(lambda p: 'tags' not in p.keys(), config))
```

Comment:
For simple statement like this, prefer list comprehension.
For more information, see [Google Python style](https://google.github.io/styleguide/pyguide.html?showone=Deprecated_Language_Features#Deprecated_Language_Features)